### DISCUSSION TITLE
Unable to cleanly register block into woocommerce-product-elements category.

### DISCUSSION THREAD

**Discussion by leewillis77** [woocommerce/woocommerce#42927] [General]
As part of a WooCommerce extension, we're trying to expose blocks from our plugin within the WooCommerce product elements category, e.g. 
![Screenshot 2023-12-18 at 17 25 46](https://github.com/woocommerce/woocommerce/assets/1097338/0f126128-a7e2-40f9-969d-d95ce5eb356a)

However, when we register our block (in PHP since I suspect that's relevant), everything works, ie the block shows up correctly, however we also receive the following warning in the JS console:

`The block "xxxx" is registered with an invalid category "woocommerce-product-elements".`

To reproduce what we're seeing you can do the following:

- Scaffold a new block plugin using `npx @wordpress/create-block  example-plugin`
- Edit `example-plugin/build/block.json` and set the 'category' to 'woocommerce-product-elements'
- Activate the newly-created "Example plugin" plugin
- Create a new page, and browse the available blocks
- You will see that the Example Plugin block is available, and shows in the correct category - ![Screenshot 2023-12-18 at 17 30 47](https://github.com/woocommerce/woocommerce/assets/1097338/0760e658-f352-4962-993f-acdaa8c9135d)
- However there will also be a warning logged to the browser console: `The block "create-block/example-plugin" is registered with an invalid category "woocommerce-product-elements".`

What's the correct way to get the block registered in that category without generating warnings?

**Comment by danieldudzic** [comment]
Hi @leewillis77,

Thank you for bringing this issue to our attention. Our investigation suggests that your custom plugin tries to register the block before Woo registers the block category. This results in an invalid warning.

This is a known issue that arises due to a race condition. We are working on enhancing the developer experience related to registering Woo blocks. Unfortunately, we do not have an estimated time for when this issue will be resolved. However, we have a workaround that should help you eliminate this warning:

1. Remove the `editorScript` from `block.json`
2. Register the block "manually" on `enqueue_block_editor_assets` hook. (this should ensure that the WooCommerce block categories got already registered).

I know that this is less than ideal, but currently, probably it's your best bet.

I hope the below example will be able to help:


```
	public function initialize() {
		// Enqueue block assets for the editor.
		add_action( 'enqueue_block_editor_assets', array( $this, 'enqueue_block_editor_assets' ) );
	}

	public function enqueue_block_editor_assets() {
		$suffix            = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '' : '.min';
		$script_path       = '/assets/dist/admin/blocks' . $suffix . '.js';
		$script_asset_path = PLUGIN_ABSPATH . 'assets/dist/admin/blocks.asset.php';
		$script_asset      = file_exists( $script_asset_path )
			? require( $script_asset_path )
			: array(
				'dependencies' => array(),
				'version'      => PLUGIN()->get_plugin_version()
			);
		$script_url        = PLUGIN()->get_plugin_url() . $script_path;

		wp_register_script(
			'plugin-admin-blocks',
			$script_url,
			$script_asset[ 'dependencies' ],
			$script_asset[ 'version' ],
			true
		);

		wp_enqueue_script( 'plugin-admin-blocks' );

	}
```

**Comment by leewillis77** [comment]
Hi; 

I've read through your comments, and had a play, but I'm afraid I can't see how the change you suggested would work? If I remove the `editorScript` definition from `block.json` then my code doesn't get built (ie I don't get the block javascript built into `js/build`), as a consequence, there's no `.asset.php` file, and no javascript to enqueue?

Contents of build folder originally:
```
./msrp-block:
block.json       index.asset.php  index.js         render.php       style-index.css

./msrp-default-block:
block.json       index.asset.php  index.js         render.php       style-index.css
```

Contents of build folder after removing `editorScript` from `block.json`:
```
./msrp-block:
block.json  render.php

./msrp-default-block:
block.json  render.php
```

**Comment by leewillis77** [comment]
In addition, I'm struggling to see where on the PHP side of things WooCommerce registers the block category?

**Comment by nerrad** [comment]
>  PHP side of things WooCommerce registers the block category?

It doesn't happen on the PHP side. It happens in the[ editor via the Blocks API](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/index.js).

@danielwrobert and @sunyatasattva, my guess here is that the WooCommerce block categories are being registered closely to when WooCommerce Blocks are being registered which means any extensions that invoke registration _after_ WooCommerce blocks will trigger the error/warning Lee is pointing out here. If this is the case, we could look at splitting out the category registration so it can be loaded earlier in the stack and see if that fixes the issue?

**Reply by danielwrobert** [reply]
Yeah, that sounds like it could work. We'll look into this!

**Reply by sunyatasattva** [reply]
Forgive my ignorance here, folks, but I'm not sure I understand. I have a few questions:

Why doesn't it happen on the PHP side? Why are we using JS to do that? I might be missing something, but it seems to me more logical to use the Blocks API version when the categories need to be enabled or disabled dynamically (which is also kinda the use-case example of the linked docs: new category is added `onClick`).

This PR (https://github.com/woocommerce/woocommerce-blocks/pull/581) mentions that we chose this route because of the rendering of the custom icon, mentioning the docs recommendation. I'm not sure if the version of the docs mentioned at the time (5 years ago) was recommending this, but if I look at the [current example](https://developer.wordpress.org/block-editor/reference-guides/filters/block-filters/#wp-blocks-updatecategory), it is using `updateCategory`, which to me makes the most sense: we register the category via PHP, and update it via the front-end to add the icon and take advantage of WordPress SVG a11y features.

Currently, from the code, it seems we are just embedding our own SVG, and not even using the `wp.primitives.SVG` to aid a11y. I'm not familiar at all with WordPress SVG (so I'm not sure what value exactly does it bring), but it seems that currently we are basically doing the worst of both world: we are registering the entire category on the front-end, and then we are using an SVG icon _as is_.

At this point, it seems to me that we'd be better off just registering the icon on the PHP side as well, just embedding the SVG.

Again, please correct me if I'm wrong and I'm missing some important assumptions, but to me the solution we should go for would be:

1. Register the category (icon-less) on the PHP side via the [`block_categories_all` filter](https://developer.wordpress.org/reference/hooks/block_categories_all/).
2. Investigate what advantages does WordPress SVG bring us and, if valuable, wrap the icon on the front-end via `wp.primitives.SVG`.
3. Update the category on the front-end side with `setCategories`.

This should solve the “race condition” problem that @leewillis77 is experiencing, as the category should be available by the time blocks are registered.

**Reply by nerrad** [reply]
Oh I was just basing my suggestion on what I saw currently in the blocks codebase. My knowledge of the APIs around that are likely out of date so if there's a better way to handle the registration without introducing any regressions definitely go for it. Whatever solves the issue Lee has brought up 😄

**Reply by sunyatasattva** [reply]
Aah, gotcha! Ok, that solves my confusion 😅 . Indeed this could be a risky change (though we have “can be added via inserter” tests for most blocks IIRC, at least addressing some potential regressions), but I think it'd make sense to approach it this way: would you agree that probably the JS way would be more suited for dynamic categories? Because that'd be my understanding if I were to architect this from scratch (indeed, that's what I did for GG at the time).

**Reply by nerrad** [reply]
Perhaps, pretty sure it's the way it is now because some APIs weren't available (or mature) when it was first implemented and it's just never been something we've revisited.

**Reply by sunyatasattva** [reply]
Alright, thanks: it makes sense, as the PR mentioned is 5 years old and there just might have not been any reason whatsoever to even look at that code. Then I can take upon me to do some further research and open a PR. 🙏

**Reply by danielwrobert** [reply]
> Investigate what advantages does WordPress SVG bring us and, if valuable, wrap the icon on the front-end via wp.primitives.SVG.

The WordPress `SVG` is a primitive component that adds the required x-browser accessibility attributes for native `SVG` elements. To my understanding, that's the key benefit of using it.

**Comment by sunyatasattva** [comment]
Heads up that, with merging of https://github.com/woocommerce/woocommerce/pull/47836, server-side registration is now happening and should solve this problem. However, as mentioned in the PR:

> [!NOTE]
> This PR is not necessarily an endorsement for 3PDs to register custom blocks under WooCommerce categories. This might be the topic of another discussion, or the one linked above.
> However, server-side registration is nevertheless the best way to go in this case.

I think we might be able to close this discussion, then. What do you think @leewillis77 ?

**Comment by leewillis77** [comment]
> Heads up that, with merging of https://github.com/woocommerce/woocommerce/pull/47836, server-side registration is now happening and should solve this problem.

Thanks. I can confirm that running with WooCommerce 'trunk' I no longer see these console warnings, so yes, this discussion can be closed I think. 

> This PR is not necessarily an endorsement for 3PDs to register custom blocks under WooCommerce categories. This might be the topic of another discussion, or the one linked above.

Doing so makes the most sense for the user I would imagine. If an extension has several blocks to add then it may. make sense for them to register their own categories, but where an extension only offers a single block it seems to be a good UX for those to sit under the WooCommerce category rather than lingering out on their own.
